### PR TITLE
TimeSeries / StateTimeline: Add support for rendering enum fields

### PR DESCRIPTION
**Pull request by leeoniya**
hot on the heels of https://github.com/grafana/grafana/pull/64059

Adds support for rendering enum fields to both the timeseries and state timeline panels. We likely will need to provide a follow-up transformation to allow users to create enum fields based on their data.

A few follow up issues to look into:
- https://github.com/grafana/grafana/issues/72073
- https://github.com/grafana/grafana/issues/72074
- https://github.com/grafana/grafana/issues/72075
- https://github.com/grafana/grafana/issues/72077

image
Screenshot 2023-07-20 at 3 46 55 PM

<details><summary>y-axis-text-enum</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 386,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "testdata",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": [
          {
            "matcher": {
              "id": "byType",
              "options": "enum"
            },
            "properties": [
              {
                "id": "custom.lineInterpolation",
                "value": "stepAfter"
              },
              {
                "id": "custom.axisColorMode",
                "value": "series"
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 15,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 3,
      "maxDataPoints": 10,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "rawFrameContent": "[\n  {\n    \"schema\": {\n      \"refId\": \"A\",\n      \"name\": \"A\",\n      \"fields\": [\n        {\n          \"name\": \"Time\",\n          \"type\": \"time\",\n          \"typeInfo\": {\n            \"frame\": \"time.Time\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Value\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Min\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Max\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Info\",\n          \"type\": \"string\",\n          \"typeInfo\": {\n            \"frame\": \"string\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"State\",\n          \"type\": \"enum\",\n          \"typeInfo\": {\n            \"frame\": \"enum\"\n          },\n          \"config\": {\n            \"type\": {\n              \"enum\": {\n                \"text\": [\n                  \"Unknown\",\n                  \"Up\",\n                  \"Down\"\n                ],\n                \"color\": [\n                  \"orange\",\n                  \"blue\",\n                  \"red\"\n                ]\n              }\n            }\n          }\n        }\n      ]\n    },\n    \"data\": {\n      \"values\": [\n        [\n          1670369720752,\n          1670373320752,\n          1670376920752,\n          1670380520752,\n          1670384120752,\n          1670387720752,\n          1670391320752,\n          1670394920752\n        ],\n        [\n          97.56700088585262,\n          97.51875596242603,\n          97.88367836664669,\n          97.64813860191066,\n          97.46344237059202,\n          97.06932713353162,\n          97.11185691118344,\n          97.4315327939052\n        ],\n        [\n          96.17381933561288,\n          96.86449303671965,\n          95.5851820937877,\n          96.94658694041154,\n          96.7433590164638,\n          96.74913605945137,\n          95.61938577725707,\n          96.24186355022567\n        ],\n        [\n          99.69898997339244,\n          98.68620205532211,\n          99.5800005905314,\n          98.2998754185445,\n          98.02873998081307,\n          99.01798295660049,\n          97.95260164072228,\n          98.24811606848421\n        ],\n        [\n          \"down\",\n          \"down\",\n          \"up\",\n          \"down\",\n          \"down\",\n          \"down\",\n          \"up\",\n          \"up\"\n        ],\n        [\n          2,\n          2,\n          1,\n          2,\n          2,\n          2,\n          1,\n          1\n        ]\n      ]\n    }\n  }\n]",
          "refId": "A",
          "scenarioId": "raw_frame"
        },
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "rawFrameContent": "[\n  {\n    \"schema\": {\n      \"refId\": \"B\",\n      \"name\": \"B\",\n      \"fields\": [\n        {\n          \"name\": \"Time\",\n          \"type\": \"time\",\n          \"typeInfo\": {\n            \"frame\": \"time.Time\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"State2\",\n          \"type\": \"enum\",\n          \"typeInfo\": {\n            \"frame\": \"enum\"\n          },\n          \"config\": {\n            \"type\": {\n              \"enum\": {\n                \"text\": [\n                  \"Left\",\n                  \"Right\"\n                ]\n              }\n            }\n          }\n        }\n      ]\n    },\n    \"data\": {\n      \"values\": [\n        [\n          1670369720752,\n          1670373320752,\n          1670376920752,\n          1670380520752,\n          1670384120752,\n          1670387720752,\n          1670391320752,\n          1670394920752\n        ],\n        [\n          0,\n          1,\n          1,\n          0,\n          0,\n          1,\n          0,\n          1\n        ]\n      ]\n    }\n  }\n]",
          "refId": "B",
          "scenarioId": "raw_frame"
        }
      ],
      "title": "Time series Enum Field Example",
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "testdata",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "fillOpacity": 70,
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineWidth": 0,
            "spanNulls": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "#EAB839",
                "value": 95
              },
              {
                "color": "#6ED0E0",
                "value": 96
              },
              {
                "color": "#EF843C",
                "value": 97
              },
              {
                "color": "#E24D42",
                "value": 98
              },
              {
                "color": "#1F78C1",
                "value": 99
              }
            ]
          }
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "B State2"
            },
            "properties": [
              {
                "id": "mappings",
                "value": [
                  {
                    "options": {
                      "0": {
                        "color": "purple",
                        "index": 1,
                        "text": "Right"
                      },
                      "1": {
                        "color": "text",
                        "index": 0,
                        "text": "Left"
                      }
                    },
                    "type": "value"
                  }
                ]
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 15,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 4,
      "maxDataPoints": 10,
      "options": {
        "alignValue": "left",
        "legend": {
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "mergeValues": true,
        "rowHeight": 0.9,
        "showValue": "auto",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "rawFrameContent": "[\n  {\n    \"schema\": {\n      \"refId\": \"A\",\n      \"name\": \"A\",\n      \"fields\": [\n        {\n          \"name\": \"Time\",\n          \"type\": \"time\",\n          \"typeInfo\": {\n            \"frame\": \"time.Time\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Value\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Min\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Max\",\n          \"type\": \"number\",\n          \"typeInfo\": {\n            \"frame\": \"float64\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"Info\",\n          \"type\": \"string\",\n          \"typeInfo\": {\n            \"frame\": \"string\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"State\",\n          \"type\": \"enum\",\n          \"typeInfo\": {\n            \"frame\": \"enum\"\n          },\n          \"config\": {\n            \"type\": {\n              \"enum\": {\n                \"text\": [\n                  \"Unknown\",\n                  \"Up\",\n                  \"Down\"\n                ],\n                \"color\": [\n                  \"orange\",\n                  \"blue\",\n                  \"red\"\n                ]\n              }\n            }\n          }\n        }\n      ]\n    },\n    \"data\": {\n      \"values\": [\n        [\n          1670369720752,\n          1670373320752,\n          1670376920752,\n          1670380520752,\n          1670384120752,\n          1670387720752,\n          1670391320752,\n          1670394920752\n        ],\n        [\n          97.56700088585262,\n          97.51875596242603,\n          97.88367836664669,\n          97.64813860191066,\n          97.46344237059202,\n          97.06932713353162,\n          97.11185691118344,\n          97.4315327939052\n        ],\n        [\n          96.17381933561288,\n          96.86449303671965,\n          95.5851820937877,\n          96.94658694041154,\n          96.7433590164638,\n          96.74913605945137,\n          95.61938577725707,\n          96.24186355022567\n        ],\n        [\n          99.69898997339244,\n          98.68620205532211,\n          99.5800005905314,\n          98.2998754185445,\n          98.02873998081307,\n          99.01798295660049,\n          97.95260164072228,\n          98.24811606848421\n        ],\n        [\n          \"down\",\n          \"down\",\n          \"up\",\n          \"down\",\n          \"down\",\n          \"down\",\n          \"up\",\n          \"up\"\n        ],\n        [\n          2,\n          2,\n          1,\n          2,\n          2,\n          2,\n          1,\n          1\n        ]\n      ]\n    }\n  }\n]",
          "refId": "A",
          "scenarioId": "raw_frame"
        },
        {
          "datasource": {
            "type": "testdata",
            "uid": "PD8C576611E62080A"
          },
          "rawFrameContent": "[\n  {\n    \"schema\": {\n      \"refId\": \"B\",\n      \"name\": \"B\",\n      \"fields\": [\n        {\n          \"name\": \"Time\",\n          \"type\": \"time\",\n          \"typeInfo\": {\n            \"frame\": \"time.Time\",\n            \"nullable\": true\n          },\n          \"config\": {}\n        },\n        {\n          \"name\": \"State2\",\n          \"type\": \"enum\",\n          \"typeInfo\": {\n            \"frame\": \"enum\"\n          },\n          \"config\": {\n            \"type\": {\n              \"enum\": {\n                \"text\": [\n                  \"Left\",\n                  \"Right\"\n                ],\n                \"color\": [\n                  \"orange\",\n                  \"yellow\"\n                ]\n              }\n            }\n          }\n        }\n      ]\n    },\n    \"data\": {\n      \"values\": [\n        [\n          1670369720752,\n          1670373320752,\n          1670376920752,\n          1670380520752,\n          1670384120752,\n          1670387720752,\n          1670391320752,\n          1670394920752\n        ],\n        [\n          0,\n          1,\n          1,\n          0,\n          0,\n          1,\n          0,\n          1\n        ]\n      ]\n    }\n  }\n]",
          "refId": "B",
          "scenarioId": "raw_frame"
        }
      ],
      "title": "State Timeline Enum Field Example",
      "type": "state-timeline"
    }
  ],
  "refresh": false,
  "revision": 1,
  "schemaVersion": 38,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "2022-12-06T23:35:20.752Z",
    "to": "2022-12-07T06:35:20.752Z"
  },
  "timepicker": {},
  "timezone": "",
  "title": "y-axis-text-enum",
  "uid": "w_w1TjK4k",
  "version": 17,
  "weekStart": ""
}
```
</details>

### CODE REVIEW

File: packages/grafana-data/src/field/displayProcessor.ts [hunk 1]
```diff
@@ -192,10 +192,18 @@ function toStringProcessor(value: unknown): DisplayValue {
...
+  cfg?: EnumFieldConfig
+): DisplayProcessor {
+  const { palette } = theme.visualization;
+
+  const seriesColor = palette[seriesIndex % palette.length];
+
   const config = {
     text: cfg?.text ?? [],
-    color: cfg?.color ?? [],
+    color: cfg?.color ?? Array(cfg?.text!.length).fill(seriesColor),
```
Comment (leeoniya) [inline packages/grafana-data/src/field/displayProcessor.ts:206]: not great?

File: public/app/plugins/panel/timeseries/utils.ts [hunk 2]
```diff
@@ -33,6 +68,10 @@ export function prepareGraphableFields(
     series = prepareTimeSeriesLong(series);
   }
 
+  if (series.some((frame) => frame.fields.some((field) => field.type === FieldType.enum))) {
```
Comment (ryantxu) [inline public/app/plugins/panel/timeseries/utils.ts:71]:
```suggestion
  if (series.filter((frame) => frame.fields.some((field) => field.type === FieldType.enum)).length > 1) {
```
maybe? we only need to do this if there are >1 enum fields right

Comment (leeoniya) [reply public/app/plugins/panel/timeseries/utils.ts:71]: yeah, i can change this to reduce()

**Review by nmarrs — APPROVED** [review]
Tested out locally and seems to work well. Any reasons not to try and include this with 9.5?

**Comment by leeoniya** [comment]
> Tested out locally and seems to work well. Any reasons not to try and include this with 9.5?

still discussing how/where colors should be assigned in `field.display()` displayProcesser (per point) vs `field.state.seriesIndex` (per field).

File: public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts [hunk 3]
```diff
@@ -107,7 +99,7 @@ export class FlameGraphDataContainer {
     // both a backward compatibility but also to allow using a simple dataFrame without enum config. This would allow
     // users to use this panel with correct query from data sources that do not return profiles natively.
     if (enumConfig) {
-      this.labelDisplayProcessor = getEnumDisplayProcessor(theme, enumConfig);
+      this.labelDisplayProcessor = getDisplayProcessor({ field: this.labelField, theme });
```
Comment (nmarrs) [inline public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:102]: not confident about this change... @aocenas can you please take a look at these changes and verify if they are ok or not? 🙏

Comment (nmarrs) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:112]: Update - seems like issue is that field type is not being set to enum (will follow up with an issue for this in PR description)

In meantime see following resources for testing locally:

Main branch
main


PR branch
branch


Debug panel file
[debug-flamegraph.txt](https://github.com/grafana/grafana/files/12114964/debug-flamegraph.txt)

File: packages/grafana-data/src/field/displayProcessor.ts [hunk 4]
```diff
@@ -116,6 +114,29 @@ export function getDisplayProcessor(options?: DisplayProcessorOptions): DisplayP
...
+      const enumIndex = +value;
+      if (config && config.type && config.type.enum) {
+        const { text: enumText, color: enumColor } = config.type.enum;
+
+        text = enumText ? enumText[enumIndex] : `${value}`;
+        // If no color specified in enum field config we will fallback to iterating through the theme palette
+        color = enumColor ? enumColor[enumIndex] : undefined;
+
+        if (color == null) {
+          const { palette } = options.theme.visualization;
```
Comment (leeoniya) [inline packages/grafana-data/src/field/displayProcessor.ts:135]: probably move this up to line 46-ish?

**Comment by leeoniya** [comment]
this is looking pretty good now :+1:

File: public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts (hunk 3 above)
Comment (nmarrs) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:112]: Follow up -> if https://github.com/grafana/grafana/pull/72078 lands we can remove [`line 111: this.labelField.type = FieldType.enum;`](https://github.com/grafana/grafana/pull/64179/files#diff-1b690a4f4dadf472d3264cd213a79819ed5a2a76857c1ee3355fcc8410777032R111)

Comment (aocenas) [reply public/app/plugins/panel/flamegraph/components/FlameGraph/dataTransform.ts:112]: Both the backend change and this one looks good to me.

**Review by aocenas — APPROVED** [review]
The flamegraph parts look good

**Review by ryantxu — APPROVED** [review]
LGTM

image

**Comment by leeoniya** [comment]
re "breaking" changes, `getEnumDisplayProcessor` was a temporary thing that was used specifically by FlameGraph panel when it lived outside the core (in phlare repo) and before the regular `getDisplayProcessor` had support for the then-alpha `enum` field type. since FlameGraph is now in core and this PR adds `enum` support to `getDisplayProcessor`, the temporary version can be removed.

image

**Pull request merged**